ignite/engine/engine.py

```diff
@@ -835,29 +861,37 @@ def _run_once_on_dataset(self) -> float:
                         break
 
                     self._fire_event(Events.DATALOADER_STOP_ITERATION)
-                    self._setup_dataloader_iter()
+                    self._maybe_terminate()
 
+                    self._setup_dataloader_iter()
                     should_exit = True
 
                     continue
 
                 self.state.iteration += 1
                 self._fire_event(Events.ITERATION_STARTED)
+                self._maybe_terminate()
```

Comment:
Doc says about `terminate_epoch`: "it terminates the current epoch after the current iteration.", but here it could terminate while iteration uncompleted.

Reply:
How would you rephrase the doc in this case ? 
I think the current code is doing what we would expect. If we call `terminate_epoch` in a handler on ITERATION_STARTED we expect engine to end the epoch after ITERATION_STARTED event and not after ITERATION_COMPLETED.